src/Compilers/CSharp/Portable/Symbols/Source/SourceOrdinaryMethodSymbol.cs

```diff
@@ -283,32 +283,36 @@ protected override void CompleteAsyncMethodChecksBetweenStartAndFinish()
             }
         }
 
-        public override ImmutableArray<TypeParameterConstraintClause> GetTypeParameterConstraintClauses(bool canIgnoreNullableContext)
+        public override ImmutableArray<TypeParameterConstraintClause> GetTypeParameterConstraintClauses(bool canUseLightweightTypeConstraintBinding)
         {
-            if (!_lazyTypeParameterConstraints.HasValue(canIgnoreNullableContext))
+            if (!_lazyTypeParameterConstraints.HasValue(canUseLightweightTypeConstraintBinding))
             {
                 var diagnostics = DiagnosticBag.GetInstance();
                 var syntax = GetSyntax();
                 var withTypeParametersBinder =
                     this.DeclaringCompilation
                     .GetBinderFactory(syntax.SyntaxTree)
                     .GetBinder(syntax.ReturnType, syntax, this);
-                var constraints = this.MakeTypeParameterConstraints(
-                    withTypeParametersBinder,
-                    TypeParameters,
-                    syntax.TypeParameterList,
-                    syntax.ConstraintClauses,
-                    canIgnoreNullableContext,
-                    diagnostics);
-                if (TypeParameterConstraintClauseExtensions.InterlockedUpdate(ref _lazyTypeParameterConstraints, constraints) &&
-                    _lazyTypeParameterConstraints.HasValue(canIgnoreNullableContext: false))
+
+                var constraints = TypeParameterConstraintClauses.Create(
+                    this.MakeTypeParameterConstraints(
+                        withTypeParametersBinder,
+                        TypeParameters,
+                        syntax.TypeParameterList,
+                        syntax.ConstraintClauses,
+                        canUseLightweightTypeConstraintBinding,
+                        diagnostics),
+                    canUseLightweightTypeConstraintBinding);
```

Comment:
>canUseLightweightTypeConstraintBinding [](start = 20, length = 38)

If `syntax.ConstraintClauses` is empty, we should pass `false` rather than `canUseLightweightTypeConstraintBinding`.

Reply:
> If syntax.ConstraintClauses is empty, we should pass false rather than canUseLightweightTypeConstraintBinding.

I don't think so. There is still a difference between empty from lightweight vs. full binding. Diagnostics should be produced from the latter, when constraints are removed to produce an empty set.

---
In reply to: [501252724](https://github.com/dotnet/roslyn/pull/48270#discussion_r501252724) [](ancestors = 501252724)

Reply:
> There is still a difference between empty from lightweight vs. full binding.

There shouldn't be a difference if `syntax.ConstraintClauses` is empty. See the `.Count == 0` case in `ConstraintsHelper.MakeTypeParameterConstraints()`. We should treat that case as fully bound. Ideally `MakeTypeParameterConstraints()` would return `TypeParameterConstraintClauses` since that method has sufficient context to make that decision.

---
In reply to: [501281392](https://github.com/dotnet/roslyn/pull/48270#discussion_r501281392) [](ancestors = 501281392,501252724)

Reply:
I see. That works. Thanks